File: tests/python/test_logger.py
import unittest
import os
import json
from datetime import datetime
from logger import log

LOG_FILE = "logs/system.log"

def get_last_log_entry():
    """Helper function to read the last log entry."""
    with open(LOG_FILE, "r") as log_file:
        lines = log_file.readlines()
    return json.loads(lines[-1]) if lines else None

class TestLogger(unittest.TestCase):
    def setUp(self):
        # clean the log file before each test
        if os.path.exists(LOG_FILE):
            os.remove(LOG_FILE)

    def test_log_info_message(self):
        log("INFO", "test_logger.py", "Test info message")
        log_entry = get_last_log_entry()

        self.assertEqual(log_entry["level"], "INFO")
        self.assertEqual(log_entry["source"], "test_logger.py")
        self.assertEqual(log_entry["message"], "Test info message")

    def test_log_error_message(self):
        log("WARNING", "test_logger.py", "Test warning message")
        log_entry = get_last_log_entry()

        self.assertEqual(log_entry["level"], "WARNING")
        self.assertEqual(log_entry["source"], "test_logger.py")
        self.assertEqual(log_entry["message"], "Test warning message")

    def test_log_error_message(self):
        log("ERROR", "test_logger.py", "Test error message")
        log_entry = get_last_log_entry()

        self.assertEqual(log_entry["level"], "ERROR")
        self.assertEqual(log_entry["source"], "test_logger.py")
        self.assertEqual(log_entry["message"], "Test error message")

if __name__ == "__main__":
    unittest.main()

File: tests/cpp/test_logger.cpp
#include "logger.hpp"
#include <gtest/gtest.h>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>

// helper function to read the last log entry from the log file
std::string getLastLogEntry(const std::string& logFilePath) {
    std::ifstream logFile(logFilePath, std::ios::in);
    std::string lastLine;
    std::string currentLine;

    while (std::getline(logFile, currentLine)) {
        lastLine = currentLine;
    }
    logFile.close();

    return lastLine;
}

// helper function to parse a log entry
struct LogEntry {
    std::string timestamp;
    std::string level;
    std::string source;
    std::string message;
};

LogEntry parseLogEntry(const std::string& logEntry) {
    std::istringstream stream(logEntry);
    LogEntry entry;

    // Parse the log line: [timestamp] [level] [source] message
    std::getline(stream, entry.timestamp, ']');
    entry.timestamp = entry.timestamp.substr(1); // Remove leading '['
    std::getline(stream, entry.level, ']');
    entry.level = entry.level.substr(2); // Remove leading " [" and trailing ']'
    std::getline(stream, entry.source, ']');
    entry.source = entry.source.substr(2); // Remove leading " [" and trailing ']'
    std::getline(stream, entry.message);  // Remaining part is the message
    entry.message = entry.message.substr(1); // Remove leading space

    return entry;
}

// test cases for logger
TEST(LoggerTest, LogsInfoMessage) {
    Logger::log("INFO", "Test info message");

    auto lastLogLine = getLastLogEntry("logs/system.log");
    auto logEntry = parseLogEntry(lastLogLine);

    EXPECT_EQ(logEntry.level, "INFO");
    EXPECT_EQ(logEntry.source, __FILE__);
    EXPECT_EQ(logEntry.message, "Test info message");
}

TEST(LoggerTest, LogsErrorMessage) {
    Logger::log("ERROR", "Test error message");

    auto lastLogLine = getLastLogEntry("logs/system.log");
    auto logEntry = parseLogEntry(lastLogLine);

    EXPECT_EQ(logEntry.level, "ERROR");
    EXPECT_EQ(logEntry.source, __FILE__);
    EXPECT_EQ(logEntry.message, "Test error message");
}

TEST(LoggerTest, LogsWarningMessage) {
    Logger::log("WARNING", "Test warning message");

    auto lastLogLine = getLastLogEntry("logs/system.log");
    auto logEntry = parseLogEntry(lastLogLine);

    EXPECT_EQ(logEntry.level, "WARNING");
    EXPECT_EQ(logEntry.source, __FILE__);
    EXPECT_EQ(logEntry.message, "Test warning message");
}